File: merge_sort.py
def merge_sort(array):
    if (len(array) > 1):
        index_to_divide = len(array) // 2
        left = array[:index_to_divide]
        right = array[index_to_divide:]

        merge_sort(left)
        merge_sort(right)

        left_elements_index_to_look_for = 0
        right_elements_index_to_look_for = 0 
        original_arr_modification_index = 0

        while (left_elements_index_to_look_for < len(left) and right_elements_index_to_look_for < len(right)):
            left_value = left[left_elements_index_to_look_for]
            right_value = right[right_elements_index_to_look_for]

            if left_value < right_value:
                array[original_arr_modification_index] = left_value
                left_elements_index_to_look_for += 1
            else:
                array[original_arr_modification_index] = right_value
                right_elements_index_to_look_for += 1
            original_arr_modification_index += 1
        
        while (left_elements_index_to_look_for < len(left)):
            array[original_arr_modification_index] = left[left_elements_index_to_look_for]
            left_elements_index_to_look_for += 1
            original_arr_modification_index += 1
        
        while (right_elements_index_to_look_for < len(right)):
            array[original_arr_modification_index] = right[right_elements_index_to_look_for]
            right_elements_index_to_look_for += 1
            original_arr_modification_index += 1
    return array

arr = [100, -5, 14, -98, 167, 8, 4, 2, 51, 86, 57, 3, 1, -200]
print(merge_sort(arr))

File: merge_sort.ts
function mergeSort(arr: (number | string)[]) {
  if (arr.length > 1) {
    const middle = Math.floor(arr.length / 2)
    const leftArr = arr.slice(0, middle)
    const rightArr = arr.slice(middle)

    mergeSort(leftArr)
    mergeSort(rightArr)

    let indexFillingArr = 0
    let indexSearchingLeftArr = 0
    let indexSearchingRightArr = 0

    while (indexSearchingLeftArr < leftArr.length && indexSearchingRightArr < rightArr.length) {
      const leftValue = leftArr[indexSearchingLeftArr]
      const rightValue = rightArr[indexSearchingRightArr]

      if (leftValue < rightValue) {
        arr[indexFillingArr] = leftValue
        indexSearchingLeftArr++
      } else {
        arr[indexFillingArr] = rightValue
        indexSearchingRightArr++
      }
      indexFillingArr++
    }

    while (indexSearchingLeftArr < leftArr.length) {
      arr[indexFillingArr] = leftArr[indexSearchingLeftArr]
      indexFillingArr++
      indexSearchingLeftArr++
    }

    while (indexSearchingRightArr < rightArr.length) {
      arr[indexFillingArr] = rightArr[indexSearchingRightArr]
      indexFillingArr++
      indexSearchingRightArr++
    }
  }
  return arr
}

const arr = [100, -5, 14, -98, 167, 8, 4, 2, 51, 86, 57, 3, 1, -200]
console.log(mergeSort(arr))

export {}
